realtek: dsa: rtl931x: Add support for HW counters

ecsv opened the pull request:
All SoC families which are supported by the counter/MIB infrastructure from #18415 (by @janh) are register based. To retrieve the current counters, it is only necessary to read data from a specific address. Such an access is possible in an atomic context (which doesn't allow sleeps).

But RTL931x has these counters hidden behind tables. To access them, it is necessary to hold a (table) mutex - making it unsuitable for atomic contexts. But `.get_stats64` is sometimes called from contexts which don't allow sleeping. Also, all calls to `rtldsa_update_port_counters` currently require a `spinlock` - making it also unsuitable for a nested mutex.

To work around this problem, the storage and retrieval code for HW counters must be modified:

* code which needs to retrieve the `rtnl_link_stats64` must get them from a cache instead of directly reading it from HW
  - access to this cache is only protected by a spinlock
* HW access must be possible with a mutex (for SoCs which need sleeping) or a spinlock (for SoC families which only use non-sleeping register access)
* a fresh retrieval of the HW counters is only done for SoC families which don't require potential sleeping code for it
* to compensate for the missing quick refresh in `.get_stats64`, SoC families with potential sleeping code are now retrieving the counters much more often.

---

The #19663 was closed without any follow up. I have now just recovered the patches and made sure that the non-RTL931x SoCs still have a similar experience for the end-user as before.

Commit:
realtek: dsa: rework MIB read locking and polling

Some Realtek SoCs do not expose MIB counters as simple registers. Instead,
retrieving counters may require blocking operations or take longer than a
normal register read. This makes the existing approach of direct reads
unsuitable. The existing approach uses spin locks which forbid sleeping
inside their context. But some hardware accesses methods (for example table
reads) might block (sleep).

To handle this, the MIB read path is redesigned with two levels of
locking:

* A global mutex protects updates of MIB data from the hardware. This is
  necessary because reads can occur both in the polling workqueue and from
  ethtool callbacks, also two user threads might call the ethtools
  callbacks. A global mutex helps to avoid parallel reads of the same
  hardware data. For table reads, this is not necessarily required because
  they are already using a table lock. But they are the reason why
  spin-locks can no longer be used (see above).

* A per-port spinlock protects the shared memory region where per-port
  counters are copied. Avoids reading of half copied values in
  .get_stats64()

As part of this change, MIB reads were removed from .get_stats64() since
that callback can be started from an atomic context and must never sleep
(block) in this context. A shared memory region is provided which will be
updated periodically by MIB workqueue and .get_stats64() will simply return
data from the shared memory.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20631
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: dsa: rtl931x: add ethtool statistics support

Add MIB data structures and table access routines for the RTL931X family.
These counters can now be exposed through the ethtool statistics interface.

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Co-developed-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20631
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Commit:
realtek: dsa: add table-based statistics infrastructure

Some Realtek SoCs such as the RTL931X store MIB counters in tables rather
than registers. Unlike register reads, table access requires programming
the table control register, setting the command field to determine read or
write, and then polling for completion. This makes it necessary to
implement a separate path for table-based statistics.

Like register-based MIBs, the table-based MIBs also come in two types: STD
and PRIV which will require slightly different implementations.

Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20631
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv committed:
realtek: dsa: Select counter lock based on sleeping behavior

On many architectures, retrieving the HW counters from the switch is not
potentially sleeping. This would potentially allow these architectures to
retrieve the most recent values from the HW when .get_stats64 is called.
But because of the global mutex (which may sleep on lock), this would no
longer be possible.

Reintroduce the per port counters lock which protects from parallel
writes+reads of the non-link_stat counters. The locking is made abstract by
using helpers which identify the correct locking mechanism based on the
used read methods of the SoC.

Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20631
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv committed:
realtek: dsa: rtl931x: Reduce HW counters polling interval

Some SoC families require table access to get the HW counters. A mutex is
required for this access - which will potentially cause a sleep in the
current context. This is not always possible with .get_stats64 because it
is also called in atomic contexts.

For these SoCs, the retrieval of the current counters in .get_stats64 is
skipped and the counters are simply retrieved a lot more often from the HW.

Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20631
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv committed:
realtek: dsa: Refresh link_stats in .get_stats64

If an architecture doesn't need to sleep for retrieving the current
statistics from the HW, it is possible to directly retrieve the last values
from the HW when .get_stats64 is called. This avoids the stale counters
with the current refresh interval of 60 seconds.

Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20631
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

janh reviewed (approved):
Looks good to me. I did a quick test on Zyxel XGS1210-12 B1 (rtl930x), and works as it did before.

Djfe commented on target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c:
```diff
@@ -1026,9 +1065,9 @@ static void rtldsa_poll_counters(struct work_struct *work)
 
 		counters = &priv->ports[i].counters;
```
you could freely remove this variable since it's not used anymore after this commit and the later introduced new locking function also retrieves `&priv->ports[i].counters.lock` for each port itself.
Also you could rename `i` to `port` for better readability but that's just a nitpick. (you also do it in your own locking function)

ecsv replied:
Ack, completely missed this while looking through Sharadanand's code (and when then adding my own stuff on top of it).

Added a new cleanup commit for the iterator variable.

ecsv committed:
realtek: dsa: Clarify statistic port iterator variable

The functions iterating through the port statistic/counter (for
initialization or polling) use the generic name "i" for the iterator. This
makes reading the actual body of the loop cumbersome because it is not
clear that various parameters of functions are about a ports.

Suggested-by: Felix Baumann <felix.bau@gmx.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20631
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv commented:
The code needed a rebase to fix some conflicts with #20640.

```diff
--- a/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c
+++ b/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c
@@ -1662,7 +1662,6 @@
 		dev_dbg(dev, "%08x: %08x\n", 0xbb000600 + i * 4, sw_r32(0x0600 + i * 4));
 }
 
-<<<<<<<
 static u64 rtldsa_931x_stat_port_table_read(int port, unsigned int mib_size,
 					    unsigned int mib_offset, bool is_pvt)
 {
@@ -1690,7 +1689,8 @@
 	rtl_table_release(r);
 
 	return ret;
-=======
+}
+
 static void rtldsa_931x_qos_set_group_selector(int port, int group)
 {
 	sw_w32_mask(RTL93XX_PORT_TBL_IDX_CTRL_IDX_MASK(port),
@@ -1762,7 +1762,6 @@
 
 	rtldsa_931x_qos_setup_default_dscp2queue_map();
 	rtldsa_931x_qos_set_scheduling_queue_weights(priv);
->>>>>>>
 }
 
 const struct rtl838x_reg rtl931x_reg = {
```

```diff
--- a/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h
+++ b/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h
@@ -1291,10 +1291,6 @@
 void rtl838x_dbgfs_init(struct rtl838x_switch_priv *priv);
 void rtl930x_dbgfs_init(struct rtl838x_switch_priv *priv);
 
-<<<<<<<
-extern int rtldsa_max_available_queue[];
-extern int rtldsa_default_queue_weights[];
-=======
 void rtldsa_counters_lock_register(struct rtl838x_switch_priv *priv, int port)
 	__acquires(&priv->ports[port].counters.lock);
 void rtldsa_counters_unlock_register(struct rtl838x_switch_priv *priv, int port)
@@ -1303,6 +1299,8 @@
 	__acquires(&priv->counters_lock);
 void rtldsa_counters_unlock_table(struct rtl838x_switch_priv *priv, int port)
 	__releases(&priv->ports[port].counters.lock);
->>>>>>>
+
+extern int rtldsa_max_available_queue[];
+extern int rtldsa_default_queue_weights[];
 
 #endif /* _RTL838X_H */
```

```diff
--- a/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h
+++ b/target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h
@@ -1311,11 +1311,9 @@
 void rtldsa_counters_unlock_table(struct rtl838x_switch_priv *priv, int port)
 	__releases(&priv->ports[port].counters.lock);
 
-<<<<<<<
+void rtldsa_update_counters_atomically(struct rtl838x_switch_priv *priv, int port);
+
 extern int rtldsa_max_available_queue[];
 extern int rtldsa_default_queue_weights[];
-=======
-void rtldsa_update_counters_atomically(struct rtl838x_switch_priv *priv, int port);
->>>>>>>
 
 #endif /* _RTL838X_H */
```

The range-diff show only changes in the context:

```
$ git range-diff 3570dee5f0409cc8bab6a95d716a76112106307e..c5cfcaf3e3d3ed05089087a5e33f48f713e6ae32 4b1c2fb270abaf4475bf66864fd6f78920f97878..e99cb0442e2dec89194fc8334c475520028b6dec
1:  7fbb1818da ! 1:  4a081cc869 realtek: dsa: rework MIB read locking and polling
    @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h: struct rtldsa
      
      struct rtl838x_port {
     @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h: struct rtl838x_switch_priv {
    - 	u16 intf_mtus[MAX_INTF_MTUS];
    - 	int intf_mtu_count[MAX_INTF_MTUS];
    + 	 */
    + 	struct rtldsa_mst *msts;
      	struct delayed_work counters_work;
     +
     +	/**
2:  23a066ace1 = 2:  bfe577fa01 realtek: dsa: add table-based statistics infrastructure
3:  f5dea6085c ! 3:  2d3f494465 realtek: dsa: rtl931x: add ethtool statistics support
    @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c: static void r
     +	return ret;
     +}
     +
    - const struct rtl838x_reg rtl931x_reg = {
    - 	.mask_port_reg_be = rtl839x_mask_port_reg_be,
    - 	.set_port_reg_be = rtl839x_set_port_reg_be,
    + static void rtldsa_931x_qos_set_group_selector(int port, int group)
    + {
    + 	sw_w32_mask(RTL93XX_PORT_TBL_IDX_CTRL_IDX_MASK(port),
     @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c: const struct rtl838x_reg rtl931x_reg = {
      	.stat_port_rst = RTL931X_STAT_PORT_RST,
      	.stat_rst = RTL931X_STAT_RST,
4:  c4accf1885 ! 4:  24966be76a realtek: dsa: Select counter lock based on sleeping behavior
    @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h: struct rtl838
     +void rtldsa_counters_unlock_table(struct rtl838x_switch_priv *priv, int port)
     +	__releases(&priv->ports[port].counters.lock);
     +
    - #endif /* _RTL838X_H */
    + extern int rtldsa_max_available_queue[];
    + extern int rtldsa_default_queue_weights[];
    + 
     
      ## target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl839x.c ##
     @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl839x.c: const struct rtl838x_reg rtl839x_reg = {
5:  8cc7800459 ! 5:  97efa85b55 realtek: dsa: Refresh link_stats in .get_stats64
    @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h: void rtldsa_c
      
     +void rtldsa_update_counters_atomically(struct rtl838x_switch_priv *priv, int port);
     +
    - #endif /* _RTL838X_H */
    + extern int rtldsa_max_available_queue[];
    + extern int rtldsa_default_queue_weights[];
    + 
     
      ## target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl839x.c ##
     @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl839x.c: const struct rtl838x_reg rtl839x_reg = {
6:  9f4123aa65 ! 6:  bdea09f224 realtek: dsa: rtl931x: Reduce HW counters polling interval
    @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.c: const struct
     
      ## target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h ##
     @@ target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl838x.h: typedef enum {
    - #define L3_EGRESS_DMACS 2048
      #define MAX_SMACS 64
    + #define DSCP_MAP_MAX 64
      
     +/* This interval needs to be short enough to prevent an undetected counter
     + * overflow. The octet counters don't need to be considered for this, because
7:  c5cfcaf3e3 = 7:  e99cb0442e realtek: dsa: Clarify statistic port iterator variable

```
